src/qml/qmlwaveformrenderer.h

```diff
@@ -0,0 +1,586 @@
+        emit axesColorChanged(m_axesColor);
+    }
+
+    const QColor& lowColor() const {
+        return m_lowColor;
+    }
+    void setLowColor(QColor color) {
+        m_lowColor = color;
+        emit lowColorChanged(m_lowColor);
+    }
+
+    const QColor& midColor() const {
+        return m_lowColor;
+    }
+    void setMidColor(QColor color) {
+        m_midColor = color;
+        emit midColorChanged(m_lowColor);
+    }
+
+    const QColor& highColor() const {
+        return m_lowColor;
+    }
+    void setHighColor(QColor color) {
+        m_highColor = color;
+        emit highColorChanged(m_lowColor);
+    }
+
+    double gainAll() const {
+        return m_gainAll;
+    }
+    void setGainAll(double gain) {
+        m_gainAll = gain;
+        emit gainAllChanged(gain);
+    }
+
+    double gainLow() const {
+        return m_gainLow;
+    }
+    void setGainLow(double gain) {
+        m_gainLow = gain;
+        emit gainLowChanged(gain);
+    }
+
+    double gainMid() const {
+        return m_gainMid;
+    }
+    void setGainMid(double gain) {
+        m_gainMid = gain;
+        emit gainMidChanged(gain);
+    }
+
+    double gainHigh() const {
+        return m_gainHigh;
+    }
+    void setGainHigh(double gain) {
+        m_gainHigh = gain;
+        emit gainHighChanged(gain);
+    }
+
+    Renderer create(WaveformWidgetRenderer* waveformWidget) const override;
+
+  signals:
+    void axesColorChanged(const QColor&);
+    void lowColorChanged(const QColor&);
+    void midColorChanged(const QColor&);
+    void highColorChanged(const QColor&);
+    void gainAllChanged(double);
+    void gainLowChanged(double);
+    void gainMidChanged(double);
+    void gainHighChanged(double);
+
+  private:
+    QColor m_axesColor;
+    QColor m_lowColor;
+    QColor m_midColor;
+    QColor m_highColor;
+
+    double m_gainAll;
+    double m_gainLow;
+    double m_gainMid;
+    double m_gainHigh;
+
+    ::WaveformRendererAbstract::PositionSource m_position{::WaveformRendererAbstract::Play};
+    allshader::WaveformRendererSignalBase::Options m_options{
+            allshader::WaveformRendererSignalBase::Option::None};
+};
+
+class QmlWaveformRendererBeat
+        : public QmlWaveformRendererFactory {
+    Q_OBJECT
+    Q_PROPERTY(QColor color READ color WRITE setColor NOTIFY colorChanged REQUIRED)
+    QML_NAMED_ELEMENT(WaveformRendererBeat)
+
+  public:
+    QmlWaveformRendererBeat();
+
+    const QColor& color() const {
+        return m_color;
+    }
+    void setColor(QColor color) {
+        m_color = color;
+        emit colorChanged(m_color);
+    }
+
+    Renderer create(WaveformWidgetRenderer* waveformWidget) const override;
+  signals:
+    void colorChanged(const QColor&);
+
+  private:
+    QColor m_color;
+    ::WaveformRendererAbstract::PositionSource m_position{::WaveformRendererAbstract::Play};
+};
+
+class QmlWaveformMarkRange : public QObject {
+    Q_OBJECT
+    Q_PROPERTY(QColor color READ color WRITE setColor)
```

Comment:
IIRC properties also need a `NOTIFY` to properly work in Qml. At least if they're supposed to support reactive bindings.

Reply:
Yeah, it's missing on few since the property/binding was not used previously, but with my yesterday commit, it should be possible to allow runtime update of those so I'll had the missing one